### PR TITLE
feat: social register

### PR DESCRIPTION
**Pull request by wangsijie**
## Summary

Register through social account.

## Linear Issue Reference

LOG-629

## Testing

- [x] send request without code to get redirect url (authorization url)
<details>
截屏2022-02-11 下午1 17 31
</details>

- [x] open authorization url in browser and redirct back with code

- [x] send request with code, an existing user try to register will be rejected with message
<details>
截屏2022-02-11 下午1 18 22
</details>

- [x] send request with code, created a new user in database
<details>
截屏2022-02-11 下午1 19 24

截屏2022-02-11 下午1 19 43
</details>

### CODE REVIEW

File: packages/core/src/lib/social.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+export const getUserInfoByCode = async (
+  connectorId: string,
+  code: string
+): Promise<SocialUserInfo> => {
+  const connector = await getConnector(connectorId);
+  const accessToken = await connector.getAccessToken(code);
+
+  return connector.getUserInfo(accessToken);
+};
```
Comment (darcyYe) [inline packages/core/src/lib/social.ts:29]: newline at the end of files is required?

**Review by darcyYe — APPROVED** [review]
LGTM, a minor fix is needed.

File: packages/phrases/src/locales/en.ts [hunk 2]
```diff
@@ -51,6 +51,7 @@ const errors = {
     email_not_exists: 'The email address has not been registered yet.',
     phone_not_exists: 'The phone number has not been registered yet.',
     identity_not_exists: 'The social account has not been registered yet.',
+    identity_exists: 'The social account has been registered.',
```
Comment (gao-sun) [inline packages/phrases/src/locales/en.ts:54]: per #221, let's add an issue to re-scope these phrases (e.g. put all `register` phrases together in an object) cc @darcyYe

**Review by gao-sun — APPROVED** [review]

File: packages/core/src/lib/social.ts (hunk 1 above)
Comment (gao-sun) [reply packages/core/src/lib/social.ts:29]: if no red sign is shown here then there's an empty line before EOF.

File: packages/core/src/lib/register.ts [hunk 3]
```diff
@@ -97,3 +98,32 @@ export const registerWithEmailAndPasscode = async (
 
   await assignRegistrationResult(ctx, provider, id);
 };
+
+export const registerWithSocial = async (
+  ctx: WithUserLogContext<Context>,
+  provider: Provider,
+  { connectorId, code }: { connectorId: string; code: string }
+) => {
+  const userInfo = await getUserInfoByCode(connectorId, code);
```
Comment (simeng-li) [inline packages/core/src/lib/register.ts:107]: nit: would it be more precise to rename it as `getUserInfoByConnectorCode`?

File: packages/core/src/lib/social.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import { getSocialConnectorInstanceById } from '@/connectors';
+import { SocialUserInfo } from '@/connectors/types';
+import RequestError from '@/errors/RequestError';
+
+const getConnector = async (connectorId: string) => {
+  try {
+    return await getSocialConnectorInstanceById(connectorId);
```
Comment (simeng-li) [inline packages/core/src/lib/social.ts:7]: nit: could use Promise.catch directly? 

```
return await getSocialConnectorInstanceById().catch(e => {
 ...
})
```

File: packages/core/src/lib/register.ts (hunk 3 above)
Comment (darcyYe) [reply packages/core/src/lib/register.ts:107]: this 'code' field is reused and represents different types of codes: in email and SMS cases, it refers to `passcode` and refers to `authorization code` in these social connector cases. not sure whether 'ConnectorCode' is a proper naming, maybe `getUserInfoByAuthorizationCode` is a more clear name.

Comment (wangsijie) [reply packages/core/src/lib/register.ts:107]: name it as `ConnectorCode` can prevent the confusing of email and SMS, so let's keep it short.

File: packages/core/src/lib/social.ts (hunk 4 above)
Comment (wangsijie) [reply packages/core/src/lib/social.ts:7]: Lint rule prefer await and catch, so I won't change it for now.

**Pull request merged**